judge/views/problem.py

```diff
@@ -201,6 +202,91 @@ def get_context_data(self, **kwargs):
+    context_object_name = 'problem'
+    http_method_names = ['options', 'post']  # This disables GET requests, even though ProblemMixin.get exists.
+    template_name = 'problem/vote-ajax.html'
+
+    def get_context_data(self, **kwargs):
+        if not self.object.vote_permission_for_user(self.request.user).can_vote():
+            raise Http404()
+
+        context = super().get_context_data(**kwargs)
+
+        try:
+            context['vote'] = ProblemPointsVote.objects.get(voter=self.request.profile, problem=self.object)
+        except ObjectDoesNotExist:
+            context['vote'] = None
+
+        context['max_possible_vote'] = settings.DMOJ_PROBLEM_MAX_USER_POINTS_VOTE
+        context['min_possible_vote'] = settings.DMOJ_PROBLEM_MIN_USER_POINTS_VOTE
+        return context
+
+    def post(self, request, *args, **kwargs):
+        problem = self.get_object()
+        if not problem.vote_permission_for_user(request.user).can_vote():
+            return HttpResponseForbidden(_('Not allowed to vote on this problem.'), content_type='text/plain')
+
+        form = ProblemPointsVoteForm(request.POST)
+        if not form.is_valid():
+            return JsonResponse(form.errors, status=400)
+
+        with transaction.atomic():
+            # Delete any pre-existing votes.
+            ProblemPointsVote.objects.filter(voter=request.profile, problem=problem).delete()
+            vote = form.save(commit=False)
+            vote.voter = request.profile
+            vote.problem = problem
+            vote.save()
+
+        return JsonResponse({'points': vote.points})
+
+
+class DeleteProblemVote(ProblemMixin, SingleObjectMixin, View):
+    http_method_names = ['options', 'post']  # This disables GET requests, even though ProblemMixin.get exists.
+
+    def post(self, request, *args, **kwargs):
+        problem = self.get_object()
+        if not problem.vote_permission_for_user(request.user).can_vote():
+            return HttpResponseForbidden(_('Not allowed to delete votes on this problem.'), content_type='text/plain')
+
+        ProblemPointsVote.objects.filter(voter=request.profile, problem=problem).delete()
+        return JsonResponse({'message': 'success'})
+
+
+class ProblemVoteStats(ProblemMixin, DetailView):
+    context_object_name = 'problem'
+    template_name = 'problem/vote-stats-ajax.html'
+
+    def get_context_data(self, **kwargs):
+        if not self.object.vote_permission_for_user(self.request.user).can_view():
+            raise Http404()
+
+        context = super().get_context_data(**kwargs)
+
+        votes = list(self.object.problem_points_votes.order_by('points').values_list('points', flat=True))
+        context['votes'] = votes
+
+        if votes:
+            context['mean'] = mean(votes)
+            context['median'] = median(votes)
+
+        context['max_possible_vote'] = settings.DMOJ_PROBLEM_MAX_USER_POINTS_VOTE
+        context['min_possible_vote'] = max(settings.DMOJ_PROBLEM_MIN_USER_POINTS_VOTE,
```

Comment:
We can get rid of this now that we assert the values make sense at `settings.py` init time.